src/cxx_supportlib/IOTools/IOUtils.cpp

```diff
@@ -767,24 +716,21 @@ pingTcpServer(const StaticString &host, unsigned int port, unsigned long long *t
 		return false;
 	}
 
-	// Try to connect the socket one last time.
+	// Now check the final connection establishment status.
 
-	try {
-		return connectToTcpServer(state);
-	} catch (const SystemException &e) {
-		if (e.code() == ECONNREFUSED) {
-			return false;
-		} else if (e.code() == EISCONN || e.code() == EINVAL) {
-			#ifdef __FreeBSD__
-				// Work around bug in FreeBSD (discovered on
-				// January 20 2013 in daemon_controller)
-				return false;
-			#else
-				throw e;
-			#endif
-		} else {
-			throw e;
-		}
+	int connectError = 0;
+	socklen_t connectErrorLen = sizeof(connectError);
+	if (getsockopt(nbcResult.first, SOL_SOCKET, SO_ERROR, &connectError, &connectErrorLen) == -1) {
+		throw SystemException("Error checking TCP socket " + host + ":" + toString(port)
+			+ " connection establishment status", connectError);
+	}
+	if (connectError == 0) {
+		return true;
+	} else if (connectError == ECONNREFUSED) {
```

Comment:
why special case `ECONNREFUSED`?

Reply:
It can theoretically happen. And the point of pingTcpServer() is to check whether the server is connectable.

Reply:
Ok but why that one separate from all other errors?

Reply:
Because pingTcpServer()'s purpose is to check whether the server is connectable. The typical use case is "loop until it is". "We established a connection" = true, "ECONNREFUSED" = "server hasn't bound on the port yet so keep looping". Everything else is an error that should cause us to abort. If we don't return false on ECONNREFUSED then that defeats the purpose of this function.